charon-ml/src/Substitute.ml

```diff
@@ -44,29 +84,16 @@ let st_substitute_visitor =
       { binder_regions; binder_value }
 
     method! visit_RVar (subst : subst) var = subst.r_subst var
-
-    method! visit_TVar (subst : subst) var =
-      match var with
-      | Free id -> subst.ty_subst id
-      | Bound _ -> failwith "bound type variable"
-
-    method! visit_CgVar (subst : subst) var =
-      match var with
-      | Free id -> subst.cg_subst id
-      | Bound _ -> failwith "bound const generic variable"
-
-    method! visit_Clause (subst : subst) var =
-      match var with
-      | Free id -> subst.tr_subst id
-      | Bound _ -> failwith "bound trait clause variable"
-
+    method! visit_TVar (subst : subst) var = subst.ty_subst var
+    method! visit_CgVar (subst : subst) var = subst.cg_subst var
+    method! visit_Clause (subst : subst) var = subst.tr_subst var
     method! visit_Self (subst : subst) = subst.tr_self
 
-    method! visit_type_var_id (_ : subst) _ =
+    method! visit_type_var_id _ _ =
       (* We should never get here because we reimplemented [visit_TypeVar] *)
       raise (Failure "Unexpected")
 
-    method! visit_const_generic_var_id (_ : subst) _ =
+    method! visit_const_generic_var_id _ _ =
       (* We should never get here because we reimplemented [visit_Var] *)
       raise (Failure "Unexpected")
```

Comment:
I think we should remove these error cases in fact, I've hit them before when substituting the whole item (which includes the `GenericParams`). We only want to substitute variables here, for example there's a `TraitClauseId` in `TraitRefKind::ParentClause` that it would make no sense to substitute, we don't want to crash there.

If we e.g. wanted to renumber all variable ids, then yeah we'd care about all the places where these ids can show up.

Reply:
Yes, makes sense.